_posts/2024/2024-02-06-about-caching-in-eo.md

```diff
@@ -0,0 +1,285 @@
+At the end, the compiler optimizes the resulting machine code and produces an object file. 
+To speed up compilation, different files of the same project might be compiled in parallel.
+3)  Then, the [Linker](https://en.wikipedia.org/wiki/Linker_(computing)) gets object files.
+The result of the linker is an executable `.exe` file.
+
+
+To speed up the build of compiled languages, [ccache](https://ccache.dev)
+and [sccache](https://github.com/mozilla/sccache) are used.    
+`ccache` uses the hash algorithm for the hashing of code at certain stages of the build.
+`ccache` uses the hash to save a code in the cache. 
+The [`ccache` hash](https://ccache.dev/manual/4.8.2.html#_common_hashed_information) is
+based on:
+* the file contents
+* the current directory of the file
+* the name of the compiler
+* the compiler’s size and modification time
+* extensions used by the compiler.
+
+Moreover, `ccache` has two types of the hashing:
+1) `Direct mode` - the hash is generated based on the source code only. 
+This mode allows to build the program faster, since the preprocessor step is skipped.
+When using this mode, the user must be sure that the external libraries, using in a project, have not changed.
+Otherwise, the project will build with errors.
+2) `Preprocessor mode` - hash is generated based on the `.cpp` file received after the preprocessor step.
+`Preprocessor mode` is slower than `direct mode`, but the project is built without errors.
+
+`Sccache`, unlike `ccache`, allows you to store cached files not only locally, but also in a cloud data storage.
+And `sccache` supports a wider range of languages, while `ccache` focuses on caching C and C++ compiler.
+
+
+### Gradle
+[Gradle](https://gradle.org) builds projects using a
+[task graph](https://docs.gradle.org/current/userguide/build_lifecycle.html) that allows for synchronous execution 
+of certain tasks.
+`Gradle` employs
+[Incremental build](https://docs.gradle.org/current/userguide/incremental_build.html#sec:how_does_it_work),
+to speed up project builds.
+For an incremental build to work, the tasks used to build the project must have specified
+source and output files.
+The provided code snippet demonstrates the implementation of a custom task in Gradle,
+showcasing how inputs and outputs are specified to enable `Incremental build`:
+```
+task myTask {
+    inputs.dir 'src/main/java/MyTask.somebody' // Specify the input directory
+    outputs.dir 'build/classes/java/main/MyTask.somebody' // Specify the output directory
+
+    doLast {
+        // Task actions go here
+        // This code will only be executed if the inputs or outputs have changed
+    }
+}
+```
+`Gradle` uses this information to determine if a task is up-to-date and needs to perform any work.
+
+To understand how `Incremental build` works, consider the following steps:
+1) Before executing a task for the first time, `Gradle` takes a 
```

Comment:
@volodya-lombrozo I used "take" because this verb is used [the Gradle documentation](https://docs.gradle.org/current/userguide/incremental_build.html#sec:how_does_it_work)